src/adaptors/http2/http2_adaptor.c

```diff
@@ -712,70 +671,70 @@ static int send_data_callback(nghttp2_session *session,
 
     //
     // local_buffs is used in the case when TLS encryption is required before sending the data out.
-    // All the http2 data is gathered into local_buffs and the local_buffs is sent to handle_outgoing_tls
+    // All the http2 data is gathered into local_buffs and the local_buffs is sent to qd_tls_encrypt_outgoing
     // where the outgoing data is encrypted.
     //
-    qd_http2_buffer_list_t    local_buffs;
+    qd_adaptor_buffer_list_t    local_buffs;
     DEQ_INIT(local_buffs);
     bool require_tls = conn->require_tls;
 
     int bytes_sent = 0; // This should not include the header length of 9.
     bool write_buffs = false;
     if (length) {
-        qd_http2_buffer_t *tail_buff = 0;
+        qd_adaptor_buffer_t *tail_buff = 0;
         if (require_tls) {
-            tail_buff = qd_http2_buffer();
-            memcpy(qd_http2_buffer_cursor(tail_buff), framehd, HTTP2_DATA_FRAME_HEADER_LENGTH);
-            qd_http2_buffer_insert(tail_buff, HTTP2_DATA_FRAME_HEADER_LENGTH);
+            tail_buff = qd_adaptor_buffer();
+            memcpy(qd_adaptor_buffer_cursor(tail_buff), framehd, HTTP2_DATA_FRAME_HEADER_LENGTH);
```

Comment:
Can you use qd_adaptor_buffer_list_append() instead of a manual memory copy?  It will do better packing of the data.